packages/wdio-cli/src/launcher.js

```diff
@@ -269,20 +269,43 @@ class Launcher {
         // If an arg appears multiple times the last occurrence is used
         let execArgv = [ ...defaultArgs, ...debugArgs, ...capExecArgs ]
 
-        // prefer launcher settings in capabilities over general launcher
-        const worker = this.runner.run({
-            cid,
-            command: 'run',
-            configFile: this.configFile,
-            argv: this.argv,
-            caps,
-            specs,
-            server,
-            execArgv
-        })
-        worker.on('message', ::this.interface.onMessage)
-        worker.on('error', ::this.interface.onMessage)
-        worker.on('exit', ::this.endHandler)
+        let retries = config.specFileRetries || 0
+        const spawn = () => {
```

Comment:
`spawn` and `buffer` need to be moved out as utilities otherwise it is impossible to unit tests these.

Reply:
@christian-bromann thanks for the feedback. As mentioned in the PR description, having the code here doesn't stop the events from every run of the same spec being sent to the reporters. Someone with more knowledge of the codebase (like yourself) should make a recommendation on the best place to put this, and also what  approach (1, 2 or 3) should be taken regarding the buffering of events.

Reply:
Tbh... you probably should still report the failed test to the reporters either way. If a test is flakey, you I would want to explicitly know about it.

Reply:
I agree @abjerstedt. At least in an ideal world where we could quickly patch all reporters to handle this correctly, this reporting every single detail would certainly be beneficial. A reporter can then simply discard what it doesn't need.
If we were to take this approach, where everything is sent to the reporters, it would cause existing reporters to report duplicates for retried specfiles, with all the first ones failed, except (possibly/hopefully) the last one. When viewed as a whole, however, this could cause the testrun as a whole to fail, since the reporter has been told of the failures. If only exit code is checked for success, this could be more easily worked around. We can also emit some kind of event that would link a failed run to its retry, which reporters can then be updated to take into account.

With this approach, I suspect, most of the code here can be reverted, and we can simply instantiate a brand new runner with the same details for a retry, and emit an event to link it to the failed run (or add a field to the runner:start event of the retry, linking back to the old run).

Thoughts?

Reply:
I like this idea. Ultimately the reporter shouldn't log anything until the test has finished which only should happen if there is no retry set or retry is set and retried that amount of time. Since we are in the early adaption face we can  do these "breaking" changes because I haven't seen a reporter that did the work to adapt for v5.

> Possible options include:

I like number 2. Again the reporter should collect the data and should reason about it when the test (worker) is done.

> I'm still finding my way about the codebase

How was it so far? I would be curious to know if the setup is understandable and the code good to read.

Thanks!

Reply:
wdio-cli is where the exit code is thrown. It seems we can grab the test before it hits that exit code and reset everything

Reply:
@christian-bromann personally I like the partitioning of modules. It organizes nicely. It does seem like there is some coupling left here and there, but this can be improved over time. I can sympathize with wanting to get the partitioning done asap and leaving smaller coupling issues for later - it must have been a herculean task.
What I would definitely recommend, which I imagine would be very beneficial for newcomers, is some sort of top-level overview of how everything fits together. For example, how the cli, the runners, the adapters, and the reporters etc all fit together. I imagine a well thought-out flow chart graphic would serve well here.

As for the desired approach here, I also think 2 is more desirable. I'll see if I can implement it in the following way:
- Revert most of the code here. Leave existing event reporting as-is. Don't buffer anything.
- When a runner fails, and it has retries left, then launch a new runner with the same configuration. Any events fired by the failed runner are left untouched, as well as any events fired by the new runner. This will cause reporters to receive duplicates for that runner.
- Link the failed runner to the new runner. Linking back from the new runner with a property in the `runner:start` event, something like `retryFor: faliedRunnerInstance`. I also think adding a property to the `runner:end` event of the failed runner can be useful to indicate that it will be retried, something like `retryPending: true`. Reporters can be updated to account for retries using these properties.

Does this sound more or less correct?

Reply:
> Does this sound more or less correct?

Yes. We would need to make additional changes in the reporters as separate PR with base to this one so that we can merge it all together eventually.